### ISSUE TITLE
HLS crashes with GHC 9.4.5 trying to add instance methods

### ISSUE THREAD

**Issue by fendor** [haskell/haskell-language-server#3607]
### Environment

Arch linux, GHC 9.4.5, HLS 2.0.0.0 

### Steps to reproduce

```haskell
module Lib (

  -- * Export list comment
  CradleProjectConfig,

) where

data CradleProjectConfig
  = NoExplicitConfig
  | ExplicitConfig FilePath

instance Eq CradleProjectConfig where

```

running code action "Add placeholders for all missing methods" crashes HLS

Haddock comment at the export list is important.

### Actual behaviour

Crashes HLS with the message:
```
haskell-language-server-9.4.5: missing match for IE:
(IEGroup
 (NoExtField)
 (1)
 (L
  { /home/hugin/Documents/haskell/hls-bug-4/Lib.hs:3:3-26 (Anchor /home/hugin/Documents/haskell/hls-bug-4/Lib.hs:3:3-26 UnchangedAnchor)
     }
  (WithHsDocIdentifiers
   (MultiLineDocString
    (HsDocStringGroup
     (1))
    (:|
     (L
      { /home/hugin/Documents/haskell/hls-bug-4/Lib.hs:3:7-26 (Anchor /home/hugin/Documents/haskell/hls-bug-4/Lib.hs:3:7-26 UnchangedAnchor)
         }
      (HsDocStringChunk
       " Export list comment"))
     []))
   [])))
CallStack (from HasCallStack):
  error, called at src/Language/Haskell/GHC/ExactPrint/ExactPrint.hs:4568:13 in ghc-exactprint-1.6.1.1-cdca7fe1b6b28a2b47f8b7064f240d76f6295482f73c55ca90fef10e3a49828e:Language.Haskell.GHC.ExactPrint.ExactPrint
 ```

**Comment by michaelpj** [comment]
Have we got an upstream issue for this? cc @alanz

**Comment by alanz** [comment]
The TLDR is that ghc-exactprint is not (currently) tested when run with the `-haddock` GHC option.
A quick check using https://github.com/ghc/ghc/blob/182df90e4d1f652c3d078294921805b9b982671b/testsuite/tests/haddock/haddock_examples/Test.hs (and `-haddock` added on line 1) shows that all the haddock comments show up twice, but are not being printed properly at present.

I will investigate further, to see if I can cause it to work by ignoring the allocated haddock comments.  A future GHC change should be to ensure that the haddock comments appear once in the AST only, at the post-processed allocated positions

**Comment by alanz** [comment]
https://gitlab.haskell.org/ghc/ghc/-/issues/23459 for proper upstream fix

**Comment by alanz** [comment]
https://hackage.haskell.org/package/ghc-exactprint-1.6.1.2

**Comment by alanz** [comment]
And I will update ghc-exactprint for GHC 9.6.x too, real soon now

**Comment by alanz** [comment]
https://github.com/alanz/ghc-exactprint/pull/127

**Comment by alanz** [comment]
https://hackage.haskell.org/package/ghc-exactprint-1.7.0.1 for GHC 9.6.X  And picked up an issue with nested, named haddock comments, will be releasing 1.6.1.3 soon

**Comment by alanz** [comment]
https://hackage.haskell.org/package/ghc-exactprint-1.6.1.3

**Comment by alanz** [comment]
And it seems the GHC 9.2.x lexer doesn't capture the comments properly, not distinguishing between

```haskell
-- | A doc comment
```

And

```haskell
{-| A doc comment -}
```

It might be possible to do gymnastics around the length of the `SrcSpan` enclosing the comment, but I would prefer not to have to do that.

Can HLS do modifications without `-haddock` (for GHC 9.2.x and before?)

**Comment by michaelpj** [comment]
I don't know if we can easily adjust whether or not we are building with `-haddock`..

**Comment by fendor** [comment]
We previously parsed the AST twice, once with `-haddock`, once without it because earlier GHC versions crashed on invalid Haddock comments.

**Comment by July541** [comment]
Compiled with ghc-exactprint-1.6.1.3 on ghc-9.4.5, it prints additional class names instead of the crash for my project, we have some extra work to do on this.

Haven't tested on the original issue code, will try to do it later.

**Comment by July541** [comment]
Class with constraints makes hls-class-plugin prints extra class names.

To produce, just replace Eq with Monoid.

`instance Monoid CradleProjectConfig where`

**Comment by michaelpj** [comment]
Is this still a problem?

**Comment by jhrcek** [comment]
Tried the reproducers (the one from description and also the Monoid mentioned later) with ghc 9.4.8 and it works fine now.

**Comment by fendor** [comment]
Thanks for checking!
